sympy/polys/matrices/domainmatrix.py

```diff
@@ -242,6 +241,18 @@ def from_Matrix(cls, M, **kwargs):
         Matrix
 
         """
+        m, n = M.shape
+        non_zero = 0
+
+        for i in range(m):
+            for j in range(n):
+                if not M[i, j].is_zero:
+                    non_zero += 1
+
+        if non_zero < (m*(n - 1) - 1)/2:
+            ddm = cls.from_list_sympy(*M.shape, M.tolist(), **kwargs)
+            return ddm.to_sparse()
```

Comment:
We should construct the sparse representation directly rather than creating a dense representation and then converting it to sparse.

Reply:
I tried doing that earlier, but all the methods for creating sparse representation required a `domain`. So I used `construct_domain` but that constructs the smallest domain containing elements of the matrix.

Reply:
@oscarbenjamin, how would you suggest I directly convert to sparse representation by using an appropriate domain.

Reply:
I guess that we need something analogous to `from_list_sympy` except that it takes a sparse form of input.

We should add a method to `Matrix` and `SparseMatrix` that can efficiently make a dict of dicts in the same format as used by SDM. Currently they have `todok` but that is a different format and also it is very slow for `Matrix` because it uses `Matrix.__getindex__` for each element of the matrix.

Reply:
Matrix should have a method like this:
```python
def todod(M):
    rowsdict = {}
    Mlol = M.tolist()
    for i, Mi in enumerate(Mlol):
        row = {j: Mij for j, Mij in enumerate(Mi) if Mij}
        if row:
            rowsdict[i] = row
    return rowsdict
```

Reply:
Ok, I'll add this.